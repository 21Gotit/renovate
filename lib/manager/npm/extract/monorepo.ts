import is from '@sindresorhus/is';
import minimatch from 'minimatch';
import upath from 'upath';
import { logger } from '../../../logger';
import { SkipReason } from '../../../types';
import type { PackageFile } from '../../types';
import { NpmManagerData } from '../types';

function matchesAnyPattern(val: string, patterns: string[]): boolean {
  const res = patterns.some(
    (pattern) => pattern === val + '/' || minimatch(val, pattern, { dot: true })
  );
  logger.trace({ val, patterns, res }, `matchesAnyPattern`);
  return res;
}

export function detectMonorepos(
  packageFiles: Partial<PackageFile>[],
  updateInternalDeps: boolean
): void {
  logger.debug('Detecting Lerna and Yarn Workspaces');
  for (const p of packageFiles) {
    const {
      packageFile,
      npmLock,
      yarnLock,
      lernaPackages,
      managerData = {},
    } = p;
    const {
      lernaClient,
      lernaDir,
      yarnWorkspacesPackages,
    } = managerData as NpmManagerData;
    const basePath = upath.dirname(packageFile);
    const packages = yarnWorkspacesPackages || lernaPackages;
    if (packages?.length) {
      logger.debug(
        { packageFile, yarnWorkspacesPackages, lernaPackages },
        'Found monorepo packages with base path ' + JSON.stringify(basePath)
      );
      const internalPackagePatterns = (is.array(packages)
        ? packages
        : [packages]
      ).map((pattern) => upath.join(basePath, pattern));
      const internalPackageFiles = packageFiles.filter((sp) =>
        matchesAnyPattern(
          upath.dirname(sp.packageFile),
          internalPackagePatterns
        )
      );
      const internalPackageNames = internalPackageFiles
        .map((sp) => sp.managerData?.packageJsonName)
        .filter(Boolean);
      if (!updateInternalDeps) {
        p.deps?.forEach((dep) => {
          if (internalPackageNames.includes(dep.depName)) {
            dep.skipReason = SkipReason.InternalPackage; // eslint-disable-line no-param-reassign
          }
        });
      }
      for (const subPackage of internalPackageFiles) {
        subPackage.managerData.lernaDir = lernaDir;
        subPackage.managerData.lernaClient = lernaClient;
        subPackage.yarnLock = subPackage.yarnLock || yarnLock;
        subPackage.npmLock = subPackage.npmLock || npmLock;
        if (subPackage.yarnLock) {
          subPackage.managerData.hasYarnWorkspaces = !!yarnWorkspacesPackages;
        }
        if (!updateInternalDeps) {
          subPackage.deps?.forEach((dep) => {
            if (internalPackageNames.includes(dep.depName)) {
              dep.skipReason = SkipReason.InternalPackage; // eslint-disable-line no-param-reassign
            }
          });
        }
      }
    }
  }
}
