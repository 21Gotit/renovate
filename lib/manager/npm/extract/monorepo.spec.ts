import { detectMonorepos } from './monorepo';

describe('manager/npm/extract', () => {
  describe('.extractPackageFile()', () => {
    it('uses lerna package settings', () => {
      const packageFiles = [
        {
          packageFile: 'package.json',
          managerData: { lernaDir: '.' },
          lernaPackages: ['packages/*'],
          packages: ['packages/*'],
          deps: [
            {
              depName: '@org/a',
            },
            {
              depName: '@org/b',
            },
            {
              depName: '@org/c',
            },
            {
              depName: 'foo',
            },
          ],
        },
        {
          packageFile: 'packages/a/package.json',
          managerData: { packageJsonName: '@org/a' },
          deps: [
            {
              depName: '@org/b',
            },
            {
              depName: '@org/c',
            },
            {
              depName: 'bar',
            },
          ],
        },
        {
          packageFile: 'packages/b/package.json',
          managerData: { packageJsonName: '@org/b' },
        },
      ] as any;
      detectMonorepos(packageFiles, false);
      expect(packageFiles).toMatchSnapshot();
      expect(packageFiles[1].managerData.lernaDir).toEqual('.');
      expect(
        packageFiles.some((packageFile) =>
          packageFile.deps?.some((dep) => dep.skipReason)
        )
      ).toBe(true);
    });
    it('updates internal packages', () => {
      const packageFiles = [
        {
          packageFile: 'package.json',
          managerData: { lernaDir: '.' },
          lernaPackages: ['packages/*'],
          packages: ['packages/*'],
          deps: [
            {
              depName: '@org/a',
            },
            {
              depName: '@org/b',
            },
            {
              depName: '@org/c',
            },
            {
              depName: 'foo',
            },
          ],
        },
        {
          packageFile: 'packages/a/package.json',
          managerData: { packageJsonName: '@org/a' },
          deps: [
            {
              depName: '@org/b',
            },
            {
              depName: '@org/c',
            },
            {
              depName: 'bar',
            },
          ],
        },
        {
          packageFile: 'packages/b/package.json',
          managerData: { packageJsonName: '@org/b' },
        },
      ] as any;
      detectMonorepos(packageFiles, true);
      expect(packageFiles).toMatchSnapshot();
      expect(packageFiles[1].managerData.lernaDir).toEqual('.');
      expect(
        packageFiles.some((packageFile) =>
          packageFile.deps?.some((dep) => dep.skipReason)
        )
      ).toBe(false);
    });
    it('uses yarn workspaces package settings with lerna', () => {
      const packageFiles = [
        {
          packageFile: 'package.json',
          lernaPackages: ['oldpackages/*'],
          lernaClient: 'yarn',
          managerData: {
            lernaDir: '.',
            yarnWorkspacesPackages: ['packages/*'],
          },
        },
        {
          packageFile: 'packages/a/package.json',
          managerData: { packageJsonName: '@org/a' },
        },
        {
          packageFile: 'packages/b/package.json',
          managerData: { packageJsonName: '@org/b' },
        },
      ];
      detectMonorepos(packageFiles, false);
      expect(packageFiles).toMatchSnapshot();
      expect(packageFiles[1].managerData.lernaDir).toEqual('.');
    });
    it('uses yarn workspaces package settings without lerna', () => {
      const packageFiles = [
        {
          packageFile: 'package.json',
          managerData: { yarnWorkspacesPackages: 'packages/*' },
        },
        {
          packageFile: 'packages/a/package.json',
          managerData: { packageJsonName: '@org/a' },
          yarnLock: 'yarn.lock',
        },
        {
          packageFile: 'packages/b/package.json',
          managerData: { packageJsonName: '@org/b' },
        },
      ];
      detectMonorepos(packageFiles, false);
      expect(packageFiles).toMatchSnapshot();
    });
  });
});
