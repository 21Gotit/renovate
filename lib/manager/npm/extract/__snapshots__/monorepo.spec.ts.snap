// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/npm/extract .extractPackageFile() updates internal packages 1`] = `
Array [
  Object {
    "deps": Array [
      Object {
        "depName": "@org/a",
      },
      Object {
        "depName": "@org/b",
      },
      Object {
        "depName": "@org/c",
      },
      Object {
        "depName": "foo",
      },
    ],
    "lernaPackages": Array [
      "packages/*",
    ],
    "managerData": Object {
      "lernaDir": ".",
    },
    "packageFile": "package.json",
    "packages": Array [
      "packages/*",
    ],
  },
  Object {
    "deps": Array [
      Object {
        "depName": "@org/b",
      },
      Object {
        "depName": "@org/c",
      },
      Object {
        "depName": "bar",
      },
    ],
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/a",
    },
    "npmLock": undefined,
    "packageFile": "packages/a/package.json",
    "yarnLock": undefined,
  },
  Object {
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/b",
    },
    "npmLock": undefined,
    "packageFile": "packages/b/package.json",
    "yarnLock": undefined,
  },
]
`;

exports[`manager/npm/extract .extractPackageFile() uses lerna package settings 1`] = `
Array [
  Object {
    "deps": Array [
      Object {
        "depName": "@org/a",
        "skipReason": "internal-package",
      },
      Object {
        "depName": "@org/b",
        "skipReason": "internal-package",
      },
      Object {
        "depName": "@org/c",
      },
      Object {
        "depName": "foo",
      },
    ],
    "lernaPackages": Array [
      "packages/*",
    ],
    "managerData": Object {
      "lernaDir": ".",
    },
    "packageFile": "package.json",
    "packages": Array [
      "packages/*",
    ],
  },
  Object {
    "deps": Array [
      Object {
        "depName": "@org/b",
        "skipReason": "internal-package",
      },
      Object {
        "depName": "@org/c",
      },
      Object {
        "depName": "bar",
      },
    ],
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/a",
    },
    "npmLock": undefined,
    "packageFile": "packages/a/package.json",
    "yarnLock": undefined,
  },
  Object {
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/b",
    },
    "npmLock": undefined,
    "packageFile": "packages/b/package.json",
    "yarnLock": undefined,
  },
]
`;

exports[`manager/npm/extract .extractPackageFile() uses yarn workspaces package settings with lerna 1`] = `
Array [
  Object {
    "lernaClient": "yarn",
    "lernaPackages": Array [
      "oldpackages/*",
    ],
    "managerData": Object {
      "lernaDir": ".",
      "yarnWorkspacesPackages": Array [
        "packages/*",
      ],
    },
    "packageFile": "package.json",
  },
  Object {
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/a",
    },
    "npmLock": undefined,
    "packageFile": "packages/a/package.json",
    "yarnLock": undefined,
  },
  Object {
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": ".",
      "packageJsonName": "@org/b",
    },
    "npmLock": undefined,
    "packageFile": "packages/b/package.json",
    "yarnLock": undefined,
  },
]
`;

exports[`manager/npm/extract .extractPackageFile() uses yarn workspaces package settings without lerna 1`] = `
Array [
  Object {
    "managerData": Object {
      "yarnWorkspacesPackages": "packages/*",
    },
    "packageFile": "package.json",
  },
  Object {
    "managerData": Object {
      "hasYarnWorkspaces": true,
      "lernaClient": undefined,
      "lernaDir": undefined,
      "packageJsonName": "@org/a",
    },
    "npmLock": undefined,
    "packageFile": "packages/a/package.json",
    "yarnLock": "yarn.lock",
  },
  Object {
    "managerData": Object {
      "lernaClient": undefined,
      "lernaDir": undefined,
      "packageJsonName": "@org/b",
    },
    "npmLock": undefined,
    "packageFile": "packages/b/package.json",
    "yarnLock": undefined,
  },
]
`;
